### ISSUE TITLE
Review (and improve) session cookie handling

### ISSUE THREAD

**Issue by M4LuZ** [lansuite/lansuite#178]
## Expected Behavior

Session cookies must be verified for correctness.
Key used for generation must be unique per installation.
In general: It must not be possible for the client to modify cookie content without this being detected by LanSuite

## Current Behavior

* Cookie validation code is unused
* encryption key is not unique per installation
* Weak algorithm used (unsalted MD5)

## Possible Solution

Implement a standard solution to fix all of the above

## Steps to Reproduce (for bugs)

Get a session cookie and see how it is generated ;)

## Context

This a security issue with yet unknown implications

**Comment by M4LuZ** [comment]
I had a read about the various available solutions and my initial thought about using JWT seems to be not without issues. Most of them boil down to the fact that the signing algorithm  can be modified by the user.
The existing frameworks would also have required additional PHP modules to be enabled (either GDM or OpenSSL).
Thus I had a look about alternatives and would now favour [PASETO](https://github.com/paragonie/paseto).

I would give it a go if nobody has a better idea.

**Comment by M4LuZ** [comment]
It seems that the user-switch functionality may have a potential flaw.
The originial user id is stored within the cookie information on the client and not in the session.
This could - in theory - allow the user to edit the userid, causing LS to change back to a different user account with potentially higher privileges.
This would only affect administrators if the user-switching functionality is only allowed for administrators, but that also has to be reviewed

**Comment by andygrunwald** [comment]
> Most of them boil down to the fact that the signing algorithm can be modified by the user.

How? The signing algorithm is also implemented on the server side. If both don't match, it is invalid.
Do you have a source for this?
I had a deeper look into JWT also and don't see those issues.

**Comment by M4LuZ** [comment]
Sorry, didn't formulate this well, have a look here for example:
https://auth0.com/blog/critical-vulnerabilities-in-json-web-token-libraries/

**Comment by M4LuZ** [comment]
Somebody also sent me this in a different context: 
https://paragonie.com/blog/2017/12/2018-guide-building-secure-php-software#things-to-avoid

> Don't use JOSE (JWT, JWS, JWE), a suite of Internet standards that codify a series of error-prone cryptographic designs that for some reason attracts a lot of evangelists in spite of the foot-bullets written into the standards.

**Comment by andygrunwald** [comment]
Personally, I think we should stick with the standard PHP cookie/session approach.
This is well known and works for ages.
Yes, the implementation in LanSuite should be rewritten/improved. But I never had or heard of the case to scale LanSuite to thousands of requests per second.
In such scale scenarios, JWT and co come into play, because they offer stateless handling.

**Comment by M4LuZ** [comment]
The idea to use JWT was not to prepare LS for stateless session handling, but to have an easy and standardized drop-in replacement for the current cookie related code that also offers additional advantages for free.
E.g. we would get the possibility to ensure client cookie validity and thus could pass authentication info between multiple LanSuite instances (for the one-installation-per-party use-cases).
I wouldn't touch the general session handling, but just improve the cookie handling as that seems to have some issues at the moment.

**Comment by M4LuZ** [comment]
Yes and no. This is about problems that currently exist with the auth cookie implementation in LanSuite, while the other ticket aims to remove it altogether and rely only on the PHP session. If we decide to remove the auth cookie, then this becomes obsolete, of course!
